src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala

```diff
@@ -199,44 +199,77 @@ private[internal] trait TypeConstraints {
+        //    necessary?
         foreach3(tvars, tparams, variances)((tvar2, tparam2, variance2) => {
-          val ok = (tparam2 != tparam) && (
-            (bound contains tparam2)
-              ||  up && (tparam2.info.bounds.lo =:= tparam.tpeHK)
-              || !up && (tparam2.info.bounds.hi =:= tparam.tpeHK)
-            )
+          val ok = (tparam2 != tparam) && ((bound contains tparam2) || {
+            val bound2 = if (up) tparam2.info.bounds.lo else tparam2.info.bounds.hi
+            bound2 contains tparam
+          })
           if (ok) {
             if (tvar2.constr.inst eq null) cyclic = true
             solveOne(tvar2, tparam2, variance2)
           }
         })
+
         if (!cyclic) {
+          def propagateBounds(tparam2: Symbol): Unit = {
+            def addBound(variance: Variance, bound: Type) = if (bound != NoType) {
+              debuglog(s"$tvar.addBound($variance, $bound.instantiateTypeParams($tparams, $tvars))")
+              if (variance.isInvariant || variance.isCovariant) tvar addHiBound bound
+              if (variance.isInvariant || variance.isContravariant) tvar addLoBound bound
+            }
+
+            def toInst(tp: Type): Type = tp match {
+              case tv: TypeVar =>
+                val inst = tv.inst
+                if (inst == null) NoType else inst
+              case _ => tp
+            }
+
+            // Try harder to propagate information we can discover from this type variable's
+            // occurence in bounds on other type variables. Also avoid some ill-kinding issues.
+            def propagate(variance: Variance, bound: Type, instantiate: => Type): Unit = bound match {
+              // Propagate bounds of the same kind: F[A, B, C[_]] <: G[A, B, C] => F <: G
+              // Undoing eta-expansion to ensure the propagated bound is correct.
+              case GenPolyType(tparams, TypeRef(_, `tparam`, targs))
+                if tparams.corresponds(targs) { (tp, ta) => ta.typeArgs.isEmpty && ta.typeSymbolDirect == tp }
+                => addBound(variance, instantiate)
+              // Propagate type constructor: F[A] <: B => F <: B.typeConstructor
+              // Propagate type arguments: F[+A, -B] <: C => A <: C.typeArgs(0) && B >: C.typeArgs(1)
+              case TypeRef(_, tsym, targs) =>
+                val inst = toInst(instantiate)
+                if(inst != NoType) {
+                  if (tsym == tparam) addBound(variance, inst.typeConstructor)
+                  foreach3(tsym.typeParams, targs, inst.typeArgs) {
+                    (tparam, targ, inst) => propagate(tparam.variance * variance, targ, inst)
+                  }
+                }
+              case _ =>
+            }
+
+            val variance = if (up) Variance.Covariant else Variance.Contravariant
+            val bound = if (up) tparam2.info.bounds.lo else tparam2.info.bounds.hi
+            propagate(variance, bound.dealias, tparam2.tpeHK.instantiateTypeParams(tparams, tvars))
```

Comment:
Why not do a `foreach2(tparams, tvars)(propagateBounds)` below, so that `tparam2.tpeHK.instantiateTypeParams(tparams, tvars)` can simply become `tvar2`? The `tpeHK` of a type parameter `tparam2` is `TypeRef(NoPrefix, tparam2, Nil)`, upon which we subst that to the corresponding tvar...

Reply:
I realize you didn't write this code, but now that we're all on this crazy ride down the rabbit hole...

Reply:
(This one is [13 years deep](https://github.com/scala/scala/commit/d5dd908#diff-559832f08ca444d5c9c85bdebb909831R101)!)

Reply:
@adriaanm I'm not quite following what you're suggesting ... `propagateBounds` takes a single argument, so it doesn't fit as an argument to `foreach2`. We do already have `tparams.foreach(propagateBounds)` and `propagateBounds` closes over `tvars`.

Reply:
I agree, also in that case `instantiate` doesn't need to be a by-name parameter.

Reply:
Ahh ... OK, got it. Trying that now.

Reply:
My commit has that refactoring already in place. The remaining bit would be to dive into higher-order args, and peel-off eta-expansion.

Reply:
Flippin' 'eck ... the solver has barely been touched in 13 years!?!

Reply:
Yup... All the more reason to be weary of users having come to depend on its precise (mal)functioning.... Experience over the years has shown that touching old code is bound to upset users. As I said before, 2.13 is focused on library changes, so we must be especially careful as we're counting on spending our "breakage budget" on the collections rework.